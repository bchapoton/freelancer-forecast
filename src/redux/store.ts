import { configureStore } from '@reduxjs/toolkit';
import parametersReducer from './slices/ParametersSlice';
import summariesReducer from './slices/SummariesSlice';
import uiReducer from './slices/UISlice';
import { StorePreloadedState } from '../helpers/TestHelper';

export const store = storeConfigurator();

export function storeConfigurator(preloadedState?: StorePreloadedState) {
    return configureStore({
        reducer: {
            parameters: parametersReducer,
            summaries: summariesReducer,
            ui: uiReducer,
        },
        preloadedState: preloadedState,
    });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
