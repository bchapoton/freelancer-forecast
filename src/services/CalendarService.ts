import moment from '../helpers/MomentHelper';

export class CalendarService {
    /**
     * Build OpenDayPerYear according to the partTimeContext, dayOffContext and closed days policies
     *
     * @param year target year
     * @param partTimeContext part time policies
     * @param dayOffContext day off policies
     * @param nonWorkingDays non-working day of the week
     */
    buildOpenDayPerYear(
        year: number,
        partTimeContext: PartTimeContext,
        dayOffContext: DayOffContext,
        nonWorkingDays: DAY[],
    ): OpenDayPerYear {
        const result: OpenDayPerYear = {
            year,
            total: 0,
            openDayPerMonths: [],
        };

        const offDaysPerMonth: number[] = this.distributeDayOffPerMonth(dayOffContext, partTimeContext, nonWorkingDays);

        for (let i = 0; i <= 11; i++) {
            const currentOpenDayTotal: number = this.calculateOpenDaysInMonth(year, i, nonWorkingDays);
            const partTimedDays: number = this.convertPartTimeToDays(currentOpenDayTotal, partTimeContext);
            const workedDays: number = partTimedDays - offDaysPerMonth[i];
            result.total += workedDays;
            result.openDayPerMonths.push({
                total: workedDays,
                month: i,
            });
        }
        return result;
    }

    /**
     * Distribute evenly the day off on each month, add the rest of the division on last month
     *
     * @param dayOffContext the day off context
     * @param partTimeContext the part time policies
     * @param nonWorkingDays the non-working days within the week
     * @return array of dispatched off days on each months, index is the month index from Date
     */
    distributeDayOffPerMonth(
        dayOffContext: DayOffContext,
        partTimeContext: PartTimeContext,
        nonWorkingDays: DAY[],
    ): number[] {
        const offDays: number = this.convertDayOffToDays(dayOffContext, partTimeContext, nonWorkingDays);
        const dayOffPerMonth: number[] = [];
        const modulo = offDays % 12;
        const offDaysPerMonth = (offDays - (offDays % 12)) / 12;
        for (let i = 0; i <= 11; i++) {
            if (modulo !== 0 && i === 11) {
                dayOffPerMonth[i] = offDaysPerMonth + modulo;
            } else dayOffPerMonth[i] = offDaysPerMonth;
        }
        return dayOffPerMonth;
    }

    /**
     * Calculate the number of opened days in the given month
     *
     * @param year target year i.e. 2022
     * @param month target month (from Date : An integer between 0 and 11 representing the months January through December.)
     * @param nonWorkingDays closed days within the week
     */
    calculateOpenDaysInMonth(year: number, month: number, nonWorkingDays: DAY[]): number {
        let result: number = 0;
        const date: Date = new Date();
        date.setFullYear(year, month, 1);
        const dayInMonth = moment(date).daysInMonth();
        for (let i = 1; i <= dayInMonth; i++) {
            if (nonWorkingDays.indexOf(date.getDay()) === -1) result++;
            date.setDate(i + 1);
        }
        return result;
    }

    /**
     * Calculate how many days in a year represents the dayOffContext
     *
     * @param dayOffContext the dayOffContext to convert.
     * @param partTimeContext the part time policies. Used for adjust the week duration
     * @param nonWorkingDays represent the non-working days within a week.
     */
    convertDayOffToDays(dayOffContext: DayOffContext, partTimeContext: PartTimeContext, nonWorkingDays: DAY[]): number {
        const openedDays: number = 7 - nonWorkingDays.length;
        let workingDay: number;
        if (partTimeContext.percentage === 100) {
            workingDay = openedDays;
        } else {
            // reduce the working day according to the part time policies
            workingDay = openedDays * (partTimeContext.percentage / 100);
        }
        return dayOffContext.value * workingDay;
    }

    /**
     * reduce full time days according to the part time context
     *
     * @param fullTimeDays number of days in full time job
     * @param partTimeContext the part time context to convert
     * @return number of days rounded
     */
    convertPartTimeToDays(fullTimeDays: number, partTimeContext: PartTimeContext): number {
        if (partTimeContext.percentage === 100) return fullTimeDays;
        else {
            const factor: number = 1 - (100 - partTimeContext.percentage) / 100;
            // add 0.4 to always round to the higher int
            return Math.round(factor * fullTimeDays + 0.4);
        }
    }

    /**
     * Displayed value of day off context
     *
     * @param dayOffContext
     */
    formatDayOffContext(dayOffContext: DayOffContext): string {
        return dayOffContext.value + ' semaines';
    }

    /**
     * Localized and join days array
     * @param days a DAY array to format
     * @param format day format from moment
     */
    formatDays(days: DAY[], format: 'dd' | 'ddd' | 'dddd' = 'ddd'): string {
        const daysLocalized: string[] = days.map<string>((day: DAY) => this.formatDay(day, format));
        return daysLocalized.join(', ');
    }

    /**
     * Localized and join days array
     * @param day the DAY to format
     * @param format day format from moment
     */
    formatDay(day: DAY, format: 'dd' | 'ddd' | 'dddd' = 'ddd'): string {
        return moment().weekday(day).format(format);
    }

    /**
     * Convert DAY number index array to DAY array
     *
     * @param indexes DAY indexes
     */
    getDaysFromNumbers(indexes: number[]): DAY[] {
        const result: DAY[] = [];
        indexes.forEach((index: number) => {
            if (DayEnumWrapper[index] !== null && DayEnumWrapper[index] !== undefined)
                result.push(DayEnumWrapper[index]);
        });
        return result;
    }
}

const calendarService: CalendarService = new CalendarService();

export default calendarService;

export enum DAY {
    SUNDAY = 0,
    MONDAY = 1,
    TUESDAY = 2,
    WEDNESDAY = 3,
    THURSDAY = 4,
    FRIDAY = 5,
    SATURDAY = 6,
}

/**
 * Workaround to map index number value to a real DAY enum value
 * don't figured out how handle properly enum manipulation in TypeScript without type issues when handle an array of it
 */
const DayEnumWrapper: DAY[] = [
    DAY.SUNDAY,
    DAY.MONDAY,
    DAY.TUESDAY,
    DAY.WEDNESDAY,
    DAY.THURSDAY,
    DAY.FRIDAY,
    DAY.SATURDAY,
];

export function getDayEnumProperKeys(): number[] {
    // typescript enum seem not expose proper keys() function
    return Object.keys(DAY)
        .filter((value) => {
            // extract only number keys, there are the number value from enum
            return !isNaN(parseInt(value));
        })
        .map((value: string) => parseInt(value));
}

export type OpenDayPerYear = {
    year: number;
    total: number;
    openDayPerMonths: OpenDayPerMonth[];
};

export type OpenDayPerMonth = {
    month: number;
    total: number;
};

export type DayOffContext = {
    value: number;
    unit: 'weeks';
};

export type PartTimeContext = {
    percentage: number;
};
