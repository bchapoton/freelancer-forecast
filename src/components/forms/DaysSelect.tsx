import * as React from 'react';
import { useCallback, useMemo, useState } from 'react';
import OutlinedInput from '@mui/material/OutlinedInput';
import InputLabel from '@mui/material/InputLabel';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import Select, { SelectChangeEvent } from '@mui/material/Select';
import calendarService, { DAY, WeekDaysArray } from '../../services/CalendarService';
import { ActionCreatorWithPayload } from '@reduxjs/toolkit';
import { useAppDispatch } from '../../redux/hooks';
import { v4 as generateUID } from 'uuid';

type DaysSelectProps = {
    dispatchAction: ActionCreatorWithPayload<DAY[]>;
    defaultValue?: number[];
    label: string;
};

export default function DaysSelect({ dispatchAction, defaultValue = [], label }: DaysSelectProps) {
    const dispatch = useAppDispatch();
    const [internalValues, setInternalValues] = useState<number[]>(defaultValue);
    const inputId: string = useMemo<string>(() => generateUID(), []);
    const onSelectHandler = useCallback(
        (event: SelectChangeEvent<number[]>) => {
            const targetValue: string | number[] = event.target.value;
            if (Array.isArray(targetValue)) {
                setInternalValues(targetValue);
                dispatch(dispatchAction(calendarService.getDaysFromNumbers(targetValue as number[])));
            }
        },
        [dispatch, dispatchAction],
    );

    return (
        <div>
            <FormControl sx={{ width: '100%', mt: '10px', mb: '10px' }}>
                <InputLabel id={inputId + '-label'} size="small">
                    {label}
                </InputLabel>
                <Select
                    labelId={inputId + '-label'}
                    id={inputId}
                    fullWidth
                    multiple
                    value={internalValues}
                    onChange={onSelectHandler}
                    input={<OutlinedInput size="small" label={label} />}
                >
                    {WeekDaysArray.map((dayIndex: number) => (
                        <MenuItem key={'day-index-' + dayIndex} value={dayIndex}>
                            {calendarService.formatDay(dayIndex, 'dddd')}
                        </MenuItem>
                    ))}
                </Select>
            </FormControl>
        </div>
    );
}
