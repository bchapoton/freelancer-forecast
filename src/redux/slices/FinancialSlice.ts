import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import parametersService, { TotalSummary, YearSummary } from '../../services/ParametersService';

export interface FinancialState {
    yearSummary: YearSummary;
}

const emptyTotals: TotalSummary = parametersService.buildEmptyTotalSummary();

const initialState: FinancialState = {
    yearSummary: { monthSummaries: [], totals: emptyTotals },
};

export const financialSlice = createSlice({
    name: 'financial',
    initialState,
    reducers: {
        setYearSummary: (state, action: PayloadAction<YearSummary>) => {
            state.yearSummary = action.payload;
        },
    },
});

export const { setYearSummary } = financialSlice.actions;

export const selectYearSummary: (state: RootState) => YearSummary = (state: RootState) => state.financial.yearSummary;

export default financialSlice.reducer;
