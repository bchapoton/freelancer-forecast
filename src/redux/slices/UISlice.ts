import { createSlice } from '@reduxjs/toolkit';
import { RootState } from '../store';

export interface UIState {
    parametersDrawerOpen: boolean;
}

const initialState: UIState = {
    parametersDrawerOpen: false,
};

export const UISlice = createSlice({
    name: 'ui',
    initialState,
    reducers: {
        toggleParametersDrawer: (state) => {
            state.parametersDrawerOpen = !state.parametersDrawerOpen;
        },
    },
});

export const { toggleParametersDrawer } = UISlice.actions;

export const isParametersDrawerOpen: (state: RootState) => boolean = (state: RootState) =>
    state.ui.parametersDrawerOpen;

export default UISlice.reducer;
