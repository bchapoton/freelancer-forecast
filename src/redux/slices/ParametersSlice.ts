import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import { TaxationFamilyContext } from '../../services/TaxesService';
import { DAY, DayOffContext, PartTimeContext } from '../../services/CalendarService';

export interface ParametersState {
    averageDailyRate: number;
    year: number;
    vatRate: number;
    taxAllowance: number;
    socialContribution: SocialContribution[];
    taxationFamilyContext: TaxationFamilyContext;
    saving: Saving;
    partTime: PartTimeContext;
    dayOffContext: DayOffContext;
    nonWorkingDays: DAY[];
}

const initialState: ParametersState = {
    averageDailyRate: 550,
    year: 2022,
    vatRate: 20,
    taxAllowance: 34,
    socialContribution: [
        { rate: 22.2, label: 'Prestations de services (bnc et bic) et vente de marchandises (bic)' },
        { rate: 0.2, label: 'Formation prof.liberale obligatoire' },
    ],
    taxationFamilyContext: {
        incomeSplittingParts: 3,
        taxableHouseholdRevenues: 24506,
        taxAllowance: 10,
    },
    saving: { mode: 'percentage', value: 20 },
    partTime: { percentage: 100 },
    dayOffContext: { unit: 'weeks', value: 5 },
    nonWorkingDays: [DAY.SATURDAY, DAY.SUNDAY],
};

export const parametersSlice = createSlice({
    name: 'parameters',
    initialState,
    reducers: {
        setAverageDailyRate: (state, action: PayloadAction<number>) => {
            state.averageDailyRate = action.payload;
        },
        setYear: (state, action: PayloadAction<number>) => {
            state.year = action.payload;
        },
        setVatRate: (state, action: PayloadAction<number>) => {
            state.vatRate = action.payload;
        },
        setDayOffContextValue: (state, action: PayloadAction<number>) => {
            state.dayOffContext.value = action.payload;
        },
        setPartTimeValue: (state, action: PayloadAction<number>) => {
            state.partTime.percentage = action.payload;
        },
        setTaxationFamilySplittingParts: (state, action: PayloadAction<number>) => {
            state.taxationFamilyContext.incomeSplittingParts = action.payload;
        },
        setTaxationFamilyTaxableHouseholdRevenues: (state, action: PayloadAction<number>) => {
            state.taxationFamilyContext.taxableHouseholdRevenues = action.payload;
        },
        setTaxationFamilyTaxAllowance: (state, action: PayloadAction<number>) => {
            state.taxationFamilyContext.taxAllowance = action.payload;
        },
        setSavingValue: (state, action: PayloadAction<number>) => {
            state.saving.value = action.payload;
        },
        setTaxAllowance: (state, action: PayloadAction<number>) => {
            state.taxAllowance = action.payload;
        },
        setNonWorkingDays: (state, action: PayloadAction<DAY[]>) => {
            state.nonWorkingDays = action.payload;
        },
    },
});

export const {
    setAverageDailyRate,
    setYear,
    setVatRate,
    setDayOffContextValue,
    setPartTimeValue,
    setTaxationFamilySplittingParts,
    setTaxationFamilyTaxableHouseholdRevenues,
    setTaxationFamilyTaxAllowance,
    setSavingValue,
    setTaxAllowance,
    setNonWorkingDays,
} = parametersSlice.actions;

export const selectParameters: (state: RootState) => ParametersState = (state: RootState) => state.parameters;

export default parametersSlice.reducer;

export type SocialContribution = {
    rate: number;
    label: string;
};

export type Saving = {
    mode: 'percentage';
    value: number;
};
