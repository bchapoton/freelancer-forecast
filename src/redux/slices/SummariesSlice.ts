import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import parametersService, { TotalSummary, YearSummary } from '../../services/ParametersService';
import { IncomeTaxesSummary } from '../../services/TaxesService';

export interface SummariesState {
    yearSummary: YearSummary;
    incomeTaxesSummary: IncomeTaxesSummary;
}

const emptyTotals: TotalSummary = parametersService.buildEmptyTotalSummary();

export const initialState: SummariesState = {
    yearSummary: { monthSummaries: [], totals: emptyTotals },
    incomeTaxesSummary: {
        taxBracketSummaries: [],
        netTaxableIncome: {
            netTaxableIncomeResult: 0,
            contractorNetTaxableIncome: 0,
            familyNetTaxableIncome: 0,
            incomeSplittingParts: 1,
        },
        total: 0,
    },
};

export const summariesSlice = createSlice({
    name: 'summaries',
    initialState,
    reducers: {
        setYearSummary: (state, action: PayloadAction<YearSummary>) => {
            state.yearSummary = action.payload;
        },
        setIncomeTaxesSummary: (state, action: PayloadAction<IncomeTaxesSummary>) => {
            state.incomeTaxesSummary = action.payload;
        },
    },
});

export const { setYearSummary, setIncomeTaxesSummary } = summariesSlice.actions;

export const selectYearSummary: (state: RootState) => YearSummary = (state: RootState) => state.summaries.yearSummary;
export const selectIncomeTaxesSummary: (state: RootState) => IncomeTaxesSummary = (state: RootState) =>
    state.summaries.incomeTaxesSummary;

export default summariesSlice.reducer;
