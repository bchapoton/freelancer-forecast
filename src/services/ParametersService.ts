import { ParametersState, Saving, SocialContribution } from '../redux/slices/ParametersSlice';
import calendarService, { OpenDayPerMonth, OpenDayPerYear } from './CalendarService';

export class ParametersService {
    /**
     * Calculate financial report from the contractor parameters
     *
     * @param parameters
     */
    buildYearSummary(parameters: ParametersState): YearSummary {
        const openDaysYear: OpenDayPerYear = calendarService.buildOpenDayPerYear(
            parameters.year,
            parameters.partTime,
            parameters.dayOffContext,
        );
        const monthSummaries: MonthSummary[] = [];
        const totalSummary: TotalSummary = this.buildEmptyTotalSummary();

        openDaysYear.openDayPerMonths.forEach((openDayPerMonth: OpenDayPerMonth) => {
            const totalDays = openDayPerMonth.total;
            const revenue = totalDays * parameters.averageDailyRate;
            const vat = this.calculateVAT(revenue, parameters.vatRate);
            const includingTax: number = revenue + vat;
            const socialContribution = this.calculateSocialContributions(revenue, parameters.socialContribution);
            const profits = revenue - socialContribution;

            const currentMonthSummary = {
                month: openDayPerMonth.month,
                totalDays: openDayPerMonth.total,
                revenue,
                vat,
                includingTax,
                socialContribution,
                profits,
            };

            monthSummaries.push(currentMonthSummary);

            totalSummary.totalDays += currentMonthSummary.totalDays;
            totalSummary.revenue += currentMonthSummary.revenue;
            totalSummary.vat += currentMonthSummary.vat;
            totalSummary.includingTax += currentMonthSummary.includingTax;
            totalSummary.socialContribution += currentMonthSummary.socialContribution;
            totalSummary.profits += currentMonthSummary.profits;
        });

        return { monthSummaries: monthSummaries, totals: totalSummary };
    }

    /**
     * Apply VAT on revenue
     * @param revenue
     * @param rate
     */
    calculateVAT(revenue: number, rate: number): number {
        return revenue * (rate / 100);
    }

    /**
     * Apply social contributions on revenue
     *
     * @param revenue
     * @param socialContributions
     */
    calculateSocialContributions(revenue: number, socialContributions: SocialContribution[]): number {
        const rate: number = this.calculateSocialContributionsRate(socialContributions);
        return revenue * (rate / 100);
    }

    /**
     * Merge SocialContribution in a single rate
     *
     * @param socialContributions
     */
    calculateSocialContributionsRate(socialContributions: SocialContribution[]): number {
        let rate: number = 0;
        socialContributions.forEach((socialContribution: SocialContribution) => (rate += socialContribution.rate));
        return rate;
    }

    buildEmptyTotalSummary(): TotalSummary {
        return {
            totalDays: 0,
            includingTax: 0,
            socialContribution: 0,
            vat: 0,
            profits: 0,
            revenue: 0,
        };
    }

    /**
     * Apply saving policies on revenue
     *
     * @param revenue base revenue
     * @param saving saving policies
     */
    calculateSaving(revenue: number, saving: Saving): number {
        if (saving.mode === 'value') return Math.max(revenue, saving.value);
        else if (saving.mode === 'percentage') return revenue * (saving.value / 100);
        else return 0;
    }
}

const parametersService: ParametersService = new ParametersService();

export default parametersService;

/**
 * Common properties in MonthSummary and TotalSummary
 */
type BaseParentSummary = {
    totalDays: number;
    revenue: number;
    vat: number;
    includingTax: number;
    socialContribution: number;
    profits: number;
};

/**
 * Financial summary on month scope
 */
export type MonthSummary = BaseParentSummary & {
    /**
     * month index from Date (0 - 11)
     */
    month: number;
};

/**
 * Total summary of a year
 */
export type TotalSummary = BaseParentSummary;

/**
 * Hold monthSummaries and reduce them into TotalSummary
 */
export type YearSummary = {
    monthSummaries: MonthSummary[];
    totals: TotalSummary;
};
