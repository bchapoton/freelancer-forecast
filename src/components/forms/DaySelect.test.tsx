import renderer from 'react-test-renderer';
import React from 'react';
import DaysSelect from './DaysSelect';
import { setNonWorkingDays } from '../../redux/slices/ParametersSlice';
import { store } from '../../redux/store';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import { Select } from '@mui/material';
import { DAY } from '../../services/CalendarService';

it('DaySelect renders correctly', () => {
    const defaultValueFromStore = store.getState().parameters.nonWorkingDays;
    const component = renderer.create(
        <Provider store={store}>
            <DaysSelect
                label="My day selector"
                dispatchAction={setNonWorkingDays}
                uuid={() => '1'}
                defaultValue={defaultValueFromStore}
            />
        </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // trigger some changes
    const newValues: DAY[] = [DAY.MONDAY];
    const instance = component.root;
    const found = instance.findByType(Select);
    renderer.act(() => {
        if (found.props.onChange) {
            found.props.onChange({ target: { value: newValues } });
        }
    });
    expect(component.toJSON()).toMatchSnapshot();
    // check if store was updated from change
    expect(store.getState().parameters.nonWorkingDays).toEqual(newValues);
});
