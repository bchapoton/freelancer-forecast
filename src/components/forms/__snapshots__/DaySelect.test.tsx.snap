// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DaySelect renders correctly 1`] = `
<div>
  <div
    className="MuiFormControl-root css-wa0nej-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      id="1-label"
      size="small"
    >
      My day selector
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="1-label 1"
        className="MuiSelect-select MuiSelect-outlined MuiSelect-multiple MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="1"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        dimanche
        , 
        samedi
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="6,0"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            My day selector
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;

exports[`DaySelect renders correctly 2`] = `
<div>
  <div
    className="MuiFormControl-root css-wa0nej-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      id="1-label"
      size="small"
    >
      My day selector
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="1-label 1"
        className="MuiSelect-select MuiSelect-outlined MuiSelect-multiple MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="1"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        lundi
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="1"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            My day selector
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
